core/include/gnuradio-4.0/CircularBuffer.hpp

```diff
@@ -459,79 +461,186 @@ class CircularBuffer
+            : _parent(other._parent),
+              _index(other._index),
+              _internalSpan(other._internalSpan) {
+        }
+
+        ConsumableInputRange& operator=(const ConsumableInputRange& other) {
+            if (this != &other) {
+                _parent = other._parent;
+                _index = other._index;
+                _internalSpan = other._internalSpan;
+            }
+            return *this;
+        }
+
+        ConsumableInputRange(ConsumableInputRange&& other) noexcept
+            : _parent(std::exchange(other._parent, nullptr))
+            , _index(std::exchange(other._index, 0))
+            , _internalSpan(std::exchange(other._internalSpan, std::span<T>{})) {
+        }
+        ConsumableInputRange& operator=(ConsumableInputRange&& other) noexcept {
+            auto tmp = std::move(other);
+            std::swap(_parent, tmp._parent);
+            std::swap(_index, tmp._index);
+            std::swap(_internalSpan, tmp._internalSpan);
+            return *this;
         }
+        ~ConsumableInputRange() = default;
+
+        [[nodiscard]] constexpr std::size_t size() const noexcept { return _internalSpan.size(); };
+        [[nodiscard]] constexpr std::size_t size_bytes() const noexcept { return size() * sizeof(T); };
+        [[nodiscard]] constexpr bool empty() const noexcept { return _internalSpan.empty(); }
+        [[nodiscard]] constexpr iterator begin() const noexcept { return _internalSpan.begin(); }
+        [[nodiscard]] constexpr iterator end() const noexcept { return _internalSpan.end(); }
+        [[nodiscard]] constexpr const T& front() const noexcept { return _internalSpan.front(); }
+        [[nodiscard]] constexpr const T& back() const noexcept { return _internalSpan.back(); }
+        [[nodiscard]] constexpr auto first(std::size_t count) const noexcept { return _internalSpan.first(count); }
+        [[nodiscard]] constexpr auto last(std::size_t count) const noexcept { return _internalSpan.last(count); }
+        [[nodiscard]] constexpr reverse_iterator rbegin() const noexcept { return _internalSpan.rbegin(); }
+        [[nodiscard]] constexpr reverse_iterator rend() const noexcept { return _internalSpan.rend(); }
+        [[nodiscard]] constexpr const T* data() const noexcept { return _internalSpan.data(); }
+        const T& operator [](std::size_t i) const noexcept  {return _parent->_buffer->_data.data()[_index + i]; }
+        const T& operator [](std::size_t i) noexcept { return _parent->_buffer->_data.data()[_index + i]; }
+        operator const std::span<const T>&() const noexcept { return _internalSpan; }
+        operator std::span<const T>&() noexcept { return _internalSpan; }
+        operator std::span<const T>&&() = delete;
+
+        template <bool strict_check = true>
+        [[nodiscard]] bool consume(std::size_t nSamples) const noexcept {
+            if (std::atomic_load_explicit(&_parent->_isRangeConsumed, std::memory_order_acquire)) {
```

Comment:
N.B. doesn't have to be atomic because this reader is (/must be) accessed (by design) always by the same thread.

Reply:
I will leave it as it is, using atomic, and add comment to check if one can remove atomic.